torch/nn/modules/transformer.py

```diff
@@ -826,3 +834,41 @@ def _get_activation_fn(activation: str) -> Callable[[Tensor], Tensor]:
         return F.gelu
 
     raise RuntimeError("activation should be relu/gelu, not {}".format(activation))
+
+
+def _detect_is_causal_mask(
+        mask: Optional[Tensor],
+        is_causal: Optional[bool],
+) -> bool:
+    """Return whether the given attention mask is causal.
+
+    Warning:
+    If ``is_causal`` is not ``None``, its value will be returned as is.  If a
+    user supplies an incorrect ``is_causal`` hint,
+
+    ``is_causal=False`` when the mask is in fact a causal attention.mask
+       may lead to reduced performance relative to what would be achievable
+       with ``is_causal=Truee``;
+    ``is_causal=True`` when the mask is in fact not a causal attention.mask
+       may lead to incorrect and unpredictable execution - in some scenarios,
+       a causal mask may be applied based on the hint, in other execution
+       scenarios the specified mask may be used.  The choice may not appear
+       to be detrministic, in that a number of factors like alignment,
+       hardware SKU, etc influence the decision whether to use a mask or
+       rely on the hint.
+    """
+    # Prevent type refinement
+    make_causal = (is_causal is True)
+
+    if is_causal is None and mask is not None:
+        # Handles mask being 3-D or 4-D; since mask should be square,
+        # this returns the correct sequence length in any case.
+        sz = mask.size(-2)
```

Comment:
Just so I understand, the initial path was written assuming a 2-D mask only, is that correct?

Reply:
That's correct.  This recognizes the common case of 2D mask.  if somebody expanded that mask to 4 dimensions, we would not catch recognize such  causal mask.  We could allocate a 4D mask and compare against it as well.  However, that would require additional cycles and memory.  

That being said,  it is important that there is no "loss" in terms of correctness as`*is_causal` is a hint that bypasses mask usage if it is True, but uses the mask as specified when False.  As such, it's a performance optimization, and there's a benefit in recognizing the common case quickly.  I have yet to encounter a 4D-expanded causal mask, and even if we were to -- it seems that the occurrence of this is infinitesimally small.

If anybody has such a 4D causal mask, the easiest outcome might be for them to set `*is_causal=True` to indicate this. (The main feature of this is to upgrade all models written before the introduction of Accelerated Transformer that pass an attention mask to take advantage of the new Flash and memory efficient kernels.)

Reply:
Update as per discussion.  We can pass in expected size for causal mask detection and only accept it as causal mask if it has the causal mask with right size for the provided input.